### PR TITLE
docs: Introduce language implementation handbook

### PR DESCRIPTION
**Pull request by RomainMuller**
## Commit Message

The document serves as a basline reference for people looking to implement support
for new languages. It provides a high-level process and recommendations to
effectively implement support for a new language, while avoiding common pitfalls.
It is meant to evolve as more languages are implemented, and complements the
*jsii specification*.

## End Commit Message

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: docs/handbooks/language-implementation.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# Language Implementation Handbook
+
+## Foreword
+
+Implementing a new language in *jsii* is not just a matter of implementing code
+generation: mapping the *[jsii type system]* to a new programming language takes
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]:
```suggestion
generation. Mapping the *[jsii type system]* to a new programming language takes
```

File: docs/handbooks/language-implementation.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+some amounts of API design, finding ways to represent APIs in the most idiomatic
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: Either prephase what API is or use a different/more appropriate term here. Maybe 'language interface design' ?

File: docs/handbooks/language-implementation.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
+# Language Implementation Handbook
+
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: Add a sentence or 2 on what this is and who it's intended for.

File: docs/handbooks/language-implementation.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+It is possible (and sometimes desirable) to start prototyping code-generation
+for the new language, as this can hlighlight implementation challenges that need
+to be discussed in the RFC document. In any case, examples of the API signatures
+that are expected to be rendered allow early feedback to be provided by possible
+future users, and still helps identify challenges.
+
+The following questions should be answered as early as possible in the process,
+to avoid surprises later on that result in significant re-engineering effort:
+
+* What does the generated APIs look like, for the typical API idioms?
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]:
```suggestion
* What do the generated APIs look like, for the typical API idioms?
```

File: docs/handbooks/language-implementation.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  - *Classes* (constructors, properties, methods, inheritance strategy, abstract
+    members, ...)
+    + Construction experience when last argument of constructor is a struct
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: This sentence is suddenly dropping too low into the implementation without context.

Or build a bit of context. "In the CDK, with the biggest adoption of JSII, follows a pattern where every the constructor of most classes has a struct as its last parameter (e.g., ...)...."

File: docs/handbooks/language-implementation.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  - *Enums*
+  - *Interfaces* and *Structs* (properties, methods, inheritance strategy,
+    implementation, ...). In particular, how are new optional properties handled
+    (those are not considered breaking change within the [jsii type system]).
+  - *Structs* (properties, inheritance strategy, implementation, ...)
+* What information is needed in order for the code-generator to produce
+  artifacts? What should the configuration block look like?
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: What's a configuration block? (half rhetorical to point out that as a reader I'm not yet familiar with this. the other half because I genuinely don't know 😉)

File: docs/handbooks/language-implementation.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+* What is the standard way to publish packages for the new language?
+  - Is there any requirements (code signature, special metadata, ...) that need
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]:
```suggestion
  - Are there any requirements (code signature, special metadata, ...) that need
```

File: docs/handbooks/language-implementation.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    to be implemented in order to publish valid packages?
+  - How are dependencies modeled? If semantic versioning  is not the norm, what
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: Link to semantic versioning

File: docs/handbooks/language-implementation.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    is the strategy to correctly represent semantic version ranges?
+* What are the toolchain and platform requirements?
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: Couple of examples (in parenthesis) from our currently implemented langs would be useful.

File: docs/handbooks/language-implementation.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+## Code Generation
+
+Before investing too much work towards the *host library* implementation, it is
+recommended to write a first version of the code generation for the new language.
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: I would rephrase to "First, implement the first version of the code generation for the new language before getting too far into the *host library* implementation."

Also explain with a sentence on why this is recommended.

File: docs/handbooks/language-implementation.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+This work happens within the [`jsii-pacmak`] package.
+
+Initially, focus on the API signatures: for example, generate code that throws
+a *not implemented* exception when called. Start by making sure a decent API is
+generated from the [`jsii-calc`] package and its dependencies, and use those to
+implement the tests form the [Standard Compliance Suite].
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]:
```suggestion
Focus, initially, on generating correct API signatures, before getting into their 
implementation. The first version may even simply throw a *not implemented*
exception when called. 

The [`jsii-calc`] package can be used a sample consuming library which uses
JSII to generate code in all target languages. Start by making sure a decent API is
generated from this package and its dependencies, and use those to
implement the tests from the [Standard Compliance Suite].
```

File: docs/handbooks/language-implementation.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+## Host library
+
+Now that we are generating "empty shell" APIs that represent the necessary
+entities to back the [Standard Compliance Suite] tests, start implementing the
+*host library* and update the code generator to make the tests pass until all
+of them pass. It's possible to start publishing artifacts before all the tests
+pass. As soon as some basic features are working, work on [Building and
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]:
```suggestion
*host library* and update the code generator until all of the tests pass. 
It's possible to publish artifacts even when tests in the suite are failing.
As soon as some basic features are working, work on [Building and
```

File: docs/handbooks/language-implementation.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+Packaging](#building-and-packaging) can start, so early feedback can be
+gathered.
+
+> :construction: A standard architecture for the *host library* has not been
+> documented yet. Upcoming language implementations should contribute to this
+> process by documenting a general architecture that should be implementable
+> in any programming languages (and thus, asbtracting away language
+> specificities).
+
+## Building and Packaging
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: Should we also mention versioning these packages as 'experiemental'?

File: docs/handbooks/language-implementation.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+The necessary toolchains should be added to he [`jsii/superchain`] Docker image,
+so that the [`jsii-pacmak`] generation can be changed to support building ready
+to publish artifacts instead of just code.
+
+Additionally, new constructs have to be added to [`aws-delivlib`] in order to
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]:
```suggestion
Additionally, [`aws-delivlib`] needs to be augumented to support publishing artifacts to the language's package repository.
```

File: docs/handbooks/language-implementation.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+## Developer Preview
+
+Once the full [Standard Compliance Suite] passes (possibly with the exception of
+certain fringy features), and the documentation covering all aspects of using
+the language bindings have been produced, the new language can be released to
+*Developer Preview*.
+
+It is recommended that new languages stay in *Developer Preview* for a minimum
+of 4 weeks, or until they have received sufficient usage to have built
```
Comment (nija-at) [inline docs/handbooks/language-implementation.md]: 4 weeks feels very low, but ok. Should be at least double that.

File: docs/handbooks/language-implementation.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+
+## Scoping & Planning
+
+The first step of most successful projects is to start by scoping work out and
+establishing a baseline plan to execute on. For contributors not yet familiar
+with *jsii*, the [specification] document is a great place to start. In
+particular, the [New Language Intake] document provides a high-level view of the
+recommended process for implementing new language support.
+
```
Comment (eladb) [inline docs/handbooks/language-implementation.md]: I would add a table here that can be used as a TOC and summarizes the recommended activities and some rough estimation of how long they take in developer days/weeks

File: docs/handbooks/language-implementation.md [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+for the new language, as this can hlighlight implementation challenges that need
+to be discussed in the RFC document. In any case, examples of the API signatures
+that are expected to be rendered allow early feedback to be provided by possible
+future users, and still helps identify challenges.
+
+The following questions should be answered as early as possible in the process,
+to avoid surprises later on that result in significant re-engineering effort:
+
+* What does the generated APIs look like, for the typical API idioms?
+  - *Classes* (constructors, properties, methods, inheritance strategy, abstract
```
Comment (eladb) [inline docs/handbooks/language-implementation.md]: I wonder if this should be an exhaustive list that can serve as a checklist/template for the RFC

File: docs/handbooks/language-implementation.md [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+* What is the standard way to publish packages for the new language?
+  - Is there any requirements (code signature, special metadata, ...) that need
+    to be implemented in order to publish valid packages?
+  - How are dependencies modeled? If semantic versioning  is not the norm, what
+    is the strategy to correctly represent semantic version ranges?
+* What are the toolchain and platform requirements?
+
+## Code Generation
+
+Before investing too much work towards the *host library* implementation, it is
```
Comment (eladb) [inline docs/handbooks/language-implementation.md]: Perhaps the overview section should provide a high level view of which components are going to be needed: 

- code generator
- runtime library
- compliance tests
- packager
- rossetta
- publisher (see the npm module jsii-release)

File: docs/handbooks/language-implementation.md (hunk 14 above)
Comment (eladb) [reply docs/handbooks/language-implementation.md]: I’ve extracted this to jsii-release (still on eladb/jsii-release)

File: docs/handbooks/language-implementation.md [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+## Building and Packaging
+
+The necessary toolchains should be added to he [`jsii/superchain`] Docker image,
+so that the [`jsii-pacmak`] generation can be changed to support building ready
+to publish artifacts instead of just code.
+
+Additionally, new constructs have to be added to [`aws-delivlib`] in order to
+support building and publishing of those new artifacts.
+
+## Documentation
```
Comment (eladb) [inline docs/handbooks/language-implementation.md]: What about the CDK reference documentation generator? I know it’s not in the jsii repo but it definitely should be!

File: docs/handbooks/language-implementation.md (hunk 6 above)
Comment (RomainMuller) [reply docs/handbooks/language-implementation.md:52]: Added a pitch to this earlier in the document.

File: docs/handbooks/language-implementation.md (hunk 15 above)
Comment (RomainMuller) [reply docs/handbooks/language-implementation.md:111]: Replaced the `or until` with `ideally until`.

File: docs/handbooks/language-implementation.md (hunk 17 above)
Comment (RomainMuller) [reply docs/handbooks/language-implementation.md:43]: I surely intend it to become that, but wanted to avoid sounding too prescriptive before I have at least 1 RFC to look up to :)

File: docs/handbooks/language-implementation.md (hunk 19 above)
Comment (RomainMuller) [reply docs/handbooks/language-implementation.md:163]: The **CDK** reference documentation generator is out-of-scope. A generic *jsii* documentation site generator would be in scope, but such a thing does not exist and I cannot reasonably set expectations that one will implement all of this from scratch as part of a new language.
